telegram/tests/test_telegram.py

```diff
@@ -0,0 +1,201 @@
+
+    def test_translate_inbound_message_no_text(self):
+        inbound_msg = {
+            'message_id': 'Message without text',
+            'chat': 'Random chat',
+            'from': self.default_user,
+        }
+
+        message = self.transport.translate_inbound_message(inbound_msg)
+        self.assertEqual('', message['content'])
+        self.assertEqual(self.bot_username, message['to_addr'])
+        self.assertEqual(self.default_user['id'], message['from_addr'])
+
+    @inlineCallbacks
+    def test_inbound_update(self):
+        default_update = json.dumps({
+            'update_id': 'update_id',
+            'message': {
+                'message_id': 'msg_id',
+                'from': self.default_user,
+                'chat': {
+                    'id': 'chat_id',
+                    'type': self.PRIVATE
+                },
+                'date': 1234,
+                'text': 'Incoming message from Telegram!',
+            }
+        })
+        res = yield self.helper.mk_request(_method='POST',
+                                           _data=default_update)
+        self.assertEqual(res.code, 200)
+
+        [msg] = yield self.helper.wait_for_dispatched_inbound(1)
+
+        expected_update = json.loads(default_update)
+        self.assertEqual(msg['to_addr'], self.bot_username)
+        self.assertEqual(msg['from_addr'], self.default_user['id'])
+        self.assertEqual(msg['content'], expected_update['message']['text'])
+        self.assertEqual(msg['transport_type'],
+                         self.transport.transport_type)
+        self.assertEqual(msg['transport_name'],
+                         self.transport.transport_name)
+
+    def query_string_to_dict(self, query_string):
+        output = {}
+        query_string = urllib.unquote_plus(query_string)
+        params = query_string.split('&')
+        for param in params:
+            [key, value] = param.split('=')
+            output[key] = value
+        return output
+
+    def test_query_string_to_dict(self):
+        query_string = 'param=one&nextparam=one+two'
+        expected_dict = {'param': 'one', 'nextparam': 'one two'}
+        self.assertEqual(expected_dict,
+                         self.query_string_to_dict(query_string))
+
+    @inlineCallbacks
+    def test_valid_outbound_message(self):
+        msg = self.helper.make_outbound(
+            content='Outbound message!',
+            to_addr=self.default_user['id'],
+            from_addr=self.bot_username,
+        )
+        d = self.helper.dispatch_outbound(msg)
+
+        req = yield self.get_next_request()
+        self.assertEqual(req.method, 'POST')
+
+        # NOTE: this is a temporary workaround, since treq doesn't let us
+        # post data in the request body, preferring instead to automatically
+        # encode any parameters as a query string
+        outbound_msg = self.query_string_to_dict(req.content.read())
```

Comment:
We should only need to do a `json.loads()` if we are giving treq the right things for a request with a json body.